ingest/ingest-projects.py

```diff
@@ -0,0 +1,374 @@
+
+            if research_areas:
+                obj.update({"researchArea": research_areas})
+
+            org = list(participant.objects(DCO.inOrganization))
+            org = org[0] if org else None
+            if org and org.label():
+                obj.update({"organization": {"uri": str(org.identifier), "name": org.label().toPython()}})
+
+            participants.append(obj)
+
+    doc.update({"participants": participants})
+
+    reporting_years = []
+    project_updates = [faux for faux in prj.objects(DCO.hasProjectUpdate) if has_type(faux, DCO.ProjectUpdate)]
+
+    if project_updates:
+        for project_update in project_updates:
+
+            reporting_year = list(project_update.objects(DCO.forReportingYear))
+            reporting_year = reporting_year[0] if reporting_year else None
+
+
+            obj = {"uri": str(reporting_year.identifier), "year": reporting_year.label().toPython()}
+
+            reporting_years.append(obj)
+
+    doc.update({"reportingYear": reporting_years})
+
+    thumbnail = get_thumbnail(prj)
+    if thumbnail:
+        doc.update({"thumbnail": thumbnail})
+
+    return doc
+
+
+def has_type(resource, type):
+    for rtype in resource.objects(RDF.type):
+        if str(rtype.identifier) == str(type):
+            return True
+    return False
+
+
+def publish(bulk, endpoint, rebuild, mapping):
+    # if configured to rebuild_index
+    # Delete and then re-create to project index (via PUT request)
+
+    index_url = endpoint + "/dco"
+
+    if rebuild:
+        requests.delete(index_url)
+        r = requests.put(index_url)
+        if r.status_code != requests.codes.ok:
+            print(r.url, r.status_code)
+            r.raise_for_status()
+
+    # push current project document mapping
+
+    mapping_url = endpoint + "/dco/project/_mapping"
+    with open(mapping) as mapping_file:
+        r = requests.put(mapping_url, data=mapping_file)
+        if r.status_code != requests.codes.ok:
+
+            # new mapping may be incompatible with previous
+            # delete current mapping and re-push
+
+            requests.delete(mapping_url)
+            r = requests.put(mapping_url, data=mapping_file)
+            if r.status_code != requests.codes.ok:
+                print(r.url, r.status_code)
+                r.raise_for_status()
+
+    # bulk import new project documents
+    bulk_import_url = endpoint + "/_bulk"
+    r = requests.post(bulk_import_url, data=bulk)
+    if r.status_code != requests.codes.ok:
+        print(r.url, r.status_code)
+        r.raise_for_status()
+
+
+def generate(threads, sparql):
+    pool = multiprocessing.Pool(threads)
+    params = [(project, sparql) for project in get_projects(endpoint=sparql)]
+    return list(itertools.chain.from_iterable(pool.starmap(process_project, params)))
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--threads', default=8, help='number of threads to use (default = 8)')
```

Comment:
I disagree with a setting of 1.  Unless we have confirmation that a setting of 1 makes a difference on endpoint performance and stability I see no reason to configure the script to take 6-8x longer to run.  If fewer threads does make a difference, I suggest testing out 2-4 threads and setting a default there if possible.